### PR TITLE
Added jack ring buffer

### PR DESCRIPTION
**Pull request by yuval-k**
Initial implementation + tests. feedback is welcome.

solves https://github.com/RustAudio/rust-jack/issues/48

### CODE REVIEW

File: src/ringbuffer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+pub struct RingBufferWriter {
+    ringbuffer_handle: std::sync::Arc<RingBuffer>,
+}
+
+unsafe impl Send for RingBufferWriter {}
+// impl !Sync for RingBufferWriter{ }
+
+
+impl RingBufferReader {
+    fn new(handle: std::sync::Arc<RingBuffer>) -> Self {
```
Comment (wmedrano) [inline src/ringbuffer/mod.rs]: Even though it is private, can you mark this unsafe, ditto for `RingBufferWriter::new`.

**Comment by yuval-k** [comment]
Will add documentation. I wanted to see if you are Ok with the ballpark of the design \ code before

File: src/ringbuffer/mod.rs (hunk 1 above)
Comment (yuval-k) [inline src/ringbuffer/mod.rs]: Do you mind sharing what do you see here that is unsafe? as I thought that using Arc got me covered.

**Comment by yuval-k** [comment]
not really sure about passing objects.. my plan is to use it for midi messages (around 3 byes) at real time speed.

File: src/ringbuffer/mod.rs (hunk 1 above)
Comment (wmedrano) [inline src/ringbuffer/mod.rs]: Would this be possible?

```rust
    let rb = Arc::new(RingBuffer::new(n).unwrap());
    let reader_a = RingBufferReader::new(rb.clone());
    let reader_b = RingBufferReader::new(rb.clone());
    let writer_a = RingBufferWriter::new(rb.clone());
    let writer_b = RingBufferWriter::new(rb.clone());
```
If so, the single reader, single writer precondition is broken.

Comment (yuval-k) [inline src/ringbuffer/mod.rs]: it isn't possible  - that's why the new is private. I read in the docs, that while reader and writer can be used in different threads, they can only be used in a single thread: "... For this to work correctly, there can only be a single reader and a single writer thread"

So i don't allow direct creation of reader and writer. also when the feature becomes stable, would want to add the "impl !Sync" to them.

p.s.: Sorry for the delay, was in vacation - should get around to adding comments this weekend.

**Comment by yuval-k** [comment]
@wmedrano please see updated PR with comments..

File: src/ringbuffer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    /// Create a reader and writer, to use the ring buffer.
+    pub fn into_reader_writer(self) -> (RingBufferReader, RingBufferWriter) {
+        let handle = std::sync::Arc::new(self);
+
+        unsafe { (RingBufferReader::new(handle.clone()), RingBufferWriter::new(handle)) }
+    }
+
+    // Re-create the ring buffer object from reader and writer. useful if you need to call reset.
+    // Not needed for deallocation, disposing of both reader and writer will deallocate buffer resources automatically.
```
Comment (wmedrano) [inline src/ringbuffer/mod.rs]: Document panic case

File: src/ringbuffer/mod.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+impl Drop for RingBuffer {
+    fn drop(&mut self) {
+        if !self.0.is_null() {
+            unsafe { j::jack_ringbuffer_free(self.0) };
+        }
+        self.0 = 0 as *mut j::jack_ringbuffer_t;
+    }
+}
+
+/// Read end of the ring buffer. Can only be used from one thread (can be a different from the write thread).
```
Comment (wmedrano) [inline src/ringbuffer/mod.rs]: can be a different -> can be different

File: src/ringbuffer/mod.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    fn into_handle(self) -> std::sync::Arc<RingBuffer> {
+        self.ringbuffer_handle
+    }
+
+
+    /// Fill a data structure with a description of the current readable data held in the ringbuffer.
+    /// This description is returned in a two slices. Two slices are needed because the data to be read may be split across the end of the ringbuffer.
+    /// The first slice represents the bytes ready to be read. if the second slice is not empty, it is the continuation of the data that ended in the first slices. For convienence, consider using peek_iter instead.
```
Comment (wmedrano) [inline src/ringbuffer/mod.rs]: 120 char limit?

Maybe rustfmt can take care of it.

**Comment by yuval-k** [comment]
@wmedrano  updated with your comments

**Pull request merged**